text/0000-api-evolution.md

```diff
@@ -0,0 +1,798 @@
+
+One very useful mechanism would be the ability to import a trait while renaming
+some of its items, e.g. `use some_mod::SomeTrait with {foo_method as bar}`. In
+particular, when methods happen to conflict across traits defined in separate
+crates, a user of the two traits could rename one of the methods out of the way.
+
+## Thoughts on possible language changes (unofficial)
+
+The following is just a quick sketch of some focused language changes that would
+help our API evolution story.
+
+**Glob semantics**
+
+As already mentioned, the fact that glob imports currently allow *no* shadowing
+is deeply problematic: in a technical sense, it means that the addition of *any*
+public item can break downstream code arbitrarily.
+
+It would be much better for API evolution (and for ergonomics and intuition) if
+explicitly-defined items trump glob imports. But this is left to a future RFC.
+
+**Globs with fine-grained control**
+
+Another useful tool for working with globs would be the ability to *exclude*
+certain items from a glob import, e.g. something like:
+
+```rust
+use some_module::{* without Foo};
+```
+
+This is especially useful for the case where multiple modules being glob
+imported happen to export items with the same name.
+
+Another possibility would be to not make it an error for two glob imports to
+bring the same name into scope, but to generate the error only at the point that
+the imported name was actually *used*. Then collisions could be resolved simply
+by adding a single explicit, shadowing import.
+
+**Default type parameters**
+
+Some of the minor changes for moving to more generic code depends on an
+interplay between defaulted type paramters and type inference, which has been
+[accepted as an RFC](https://github.com/rust-lang/rfcs/pull/213) but not yet
+implemented.
+
+**"Extensible" enums**
+
+There is already [an RFC](https://github.com/rust-lang/rfcs/pull/757) for an
+`enum` annotation that would make it possible to add variants without ever
+breaking downstream code.
+
+**Sealed traits**
+
+The ability to annotate a trait with some "sealed" marker, saying that no
+external implementations are allowed, would be useful in certain cases where a
+crate wishes to define a closed set of types that implements a particular
+interface. Such an attribute would make it possible to evolve the interface
+without a major version bump (since no downstream implementors can exist).
+
+**Defaulted parameters**
+
+Also known as "optional arguments" -- an
+[oft-requested](https://github.com/rust-lang/rfcs/issues/323) feature. Allowing
+arguments to a function to be optional makes it possible to add new arguments
+after the fact without a major version bump.
+
+**Open-ended explicit type paramters**
+
+One hazard is that with today's explicit type parameter syntax, you must always
+specify *all* type parameters: `foo::<T, U>(x, y)`. That means that adding a new
+type parameter to `foo` can break code, even if a default is provided.
```

Comment:
default parameters for functions & methods aren't fully implemented, they don't seem to work really.